File: concept/type/EntityType.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.vaticle.typedb.core.concept.type;

import com.vaticle.typedb.core.common.iterator.sorted.SortedIterator.Forwardable;
import com.vaticle.typedb.core.common.parameters.Order;
import com.vaticle.typedb.core.common.parameters.Concept.Existence;
import com.vaticle.typedb.core.common.parameters.Concept.Transitivity;
import com.vaticle.typedb.core.concept.thing.Entity;

public interface EntityType extends ThingType {

    @Override
    EntityType getSupertype();

    @Override
    Forwardable<? extends EntityType, Order.Asc> getSupertypes();

    @Override
    Forwardable<? extends EntityType, Order.Asc> getSubtypes();

    @Override
    Forwardable<? extends EntityType, Order.Asc> getSubtypes(Transitivity transitivity);

    @Override
    Forwardable<? extends Entity, Order.Asc> getInstances();

    @Override
    Forwardable<? extends Entity, Order.Asc> getInstances(Transitivity transitivity);

    void setSupertype(EntityType superType);

    Entity create();

    Entity create(Existence existence);
}

File: tests/behaviour/concept/type/entity_type.rs
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

#![allow(unexpected_cfgs, reason = "features defined in Bazel targets aren't currently communicated to Cargo")]

use steps::Context;

#[tokio::test]
async fn test() {
    // Bazel specific path: when running the test in bazel, the external data from
    // @typedb_behaviour is stored in a directory that is a sibling to
    // the working directory.
    #[cfg(feature = "bazel")]
    let path = "../typedb_behaviour/concept/type/entitytype.feature";

    #[cfg(not(feature = "bazel"))]
    let path = "bazel-typedb/external/typedb_behaviour/concept/type/entitytype.feature";

    assert!(Context::test(path, true).await);
}
